src/System.IO.Pipelines/System/IO/Pipelines/MemoryEnumerator.cs

```diff
@@ -8,46 +8,114 @@ namespace System.IO.Pipelines
     /// </summary>
     public struct BufferEnumerator
     {
-        private SegmentEnumerator _segmentEnumerator;
-        private Memory<byte> _current;
+        private object _segment;
+        private object _cursorSegment;
+        private int _cursorStart;
+        private int _startIndex;
+        private readonly int _endIndex;
+        private readonly object _endSegment;
 
         /// <summary>
-        /// 
+        ///
         /// </summary>
         public BufferEnumerator(ReadCursor start, ReadCursor end)
         {
-            _segmentEnumerator = new SegmentEnumerator(start, end);
-            _current = default;
+            _startIndex = start.Index;
+            _segment = start.Segment;
+            _cursorSegment = null;
+            _cursorStart = 0;
+            _endSegment = end.Segment;
+            _endIndex = end.Index;
+            Current = default;
         }
 
         /// <summary>
         /// The current <see cref="Buffer{Byte}"/>
         /// </summary>
-        public Memory<byte> Current => _current;
+        public Memory<byte> Current { get; set; }
 
         /// <summary>
         /// Moves to the next <see cref="Buffer{Byte}"/> in the <see cref="ReadableBuffer"/>
         /// </summary>
         /// <returns></returns>
         public bool MoveNext()
         {
-            if (!_segmentEnumerator.MoveNext())
+            var segment = _segment;
+
+            if (segment == null)
             {
-                _current = default;
                 return false;
             }
-            var current = _segmentEnumerator.Current;
-            _current = current.Segment.Buffer.Slice(current.Start, current.Length);
 
-            return true;
+            if (segment is BufferSegment bufferSegment)
+            {
+                var start = _startIndex;
+                var end = bufferSegment.End;
+
+                if (segment == _endSegment)
+                {
+                    end = _endIndex;
+                    _segment = null;
+                }
+                else
+                {
+                    _segment = bufferSegment.Next;
+                    if (_segment == null)
+                    {
+                        if (_endSegment != null)
+                        {
+                            ThrowEndNotSeen();
+                        }
+                    }
+                    else
+                    {
+                        _startIndex = bufferSegment.Next.Start;
+                    }
+                }
+
+                Current = bufferSegment.Memory.Slice(start, end - start);
+                _cursorSegment = bufferSegment;
+                _cursorStart = start;
+                return true;
+            }
+
+            if (segment is byte[] array)
+            {
+                Current = ((Memory<byte>)array).Slice(_startIndex, _endIndex - _startIndex);
```

Comment:
Probably slightly more efficient to do `Current = new Memory<byte>(array, _startIndex, _endIndex - _startIndex)` so you only create one Memory object instead of two.